File: app/src/main/java/igrek/todotree/resources/UserInfoService.java
package igrek.todotree.resources;

import android.app.Activity;
import android.graphics.Color;
import android.support.design.widget.Snackbar;
import android.support.v4.content.ContextCompat;
import android.view.View;

import java.util.HashMap;

import javax.inject.Inject;

import igrek.todotree.R;
import igrek.todotree.dagger.DaggerIOC;
import igrek.todotree.gui.GUI;
import igrek.todotree.logger.Logs;

public class UserInfoService {
	
	@Inject
	Activity activity;
	
	@Inject
	GUI gui;
	
	private View mainView = null;
	
	private HashMap<View, Snackbar> infobars = new HashMap<>();
	
	public UserInfoService() {
		DaggerIOC.getAppComponent().inject(this);
	}
	
	public String resString(int resourceId) {
		return activity.getResources().getString(resourceId);
	}
	
	public void setMainView(View mainView) {
		this.mainView = mainView;
		infobars.clear();
	}
	
	/**
	 * @param info       tekst do wyświetlenia lub zmiany
	 * @param view       widok, na którym ma zostać wyświetlony tekst
	 * @param actionName tekst przycisku akcji (jeśli null - brak przycisku akcji)
	 * @param action     akcja kliknięcia przycisku (jeśli null - schowanie wyświetlanego tekstu)
	 */
	public void showActionInfo(String info, View view, String actionName, InfoBarClickAction action, Integer color) {
		
		if (view == null) {
			view = mainView;
		}
		
		Snackbar snackbar = infobars.get(view);
		if (snackbar == null || !snackbar.isShown()) { //nowy
			snackbar = Snackbar.make(view, info, Snackbar.LENGTH_SHORT);
			snackbar.setActionTextColor(Color.WHITE);
		} else { //widoczny - użyty kolejny raz
			snackbar.setText(info);
		}
		
		if (actionName != null) {
			if (action == null) {
				final Snackbar finalSnackbar = snackbar;
				action = new InfoBarClickAction() {
					@Override
					public void onClick() {
						finalSnackbar.dismiss();
					}
				};
			}
			
			final InfoBarClickAction finalAction = action;
			snackbar.setAction(actionName, new View.OnClickListener() {
				@Override
				public void onClick(View v) {
					finalAction.onClick();
				}
			});
			if (color != null) {
				snackbar.setActionTextColor(color);
			}
		}
		
		snackbar.show();
		infobars.put(view, snackbar);
		Logs.info(info);
	}
	
	public void showActionInfo(int resourceId, View view, String actionName, InfoBarClickAction action, Integer color) {
		showActionInfo(resString(resourceId), view, actionName, action, color);
	}
	
	public void hideInfo(View view) {
		final Snackbar snackbar = infobars.get(view);
		if (snackbar != null) {
			snackbar.dismiss();
		}
	}
	
	
	public void showInfo(String info) {
		showActionInfo(info, gui.getMainContent(), "OK", null, null);
	}
	
	public void showInfoCancellable(String info, InfoBarClickAction cancelCallback) {
		showActionInfo(info, gui.getMainContent(), "Cofnij", cancelCallback, ContextCompat.getColor(activity, R.color.colorPrimary));
	}
}

File: app/src/main/java/igrek/todotree/service/resources/UserInfoService.kt
package igrek.todotree.service.resources

import android.app.Activity
import android.graphics.Color
import android.view.View
import android.widget.Toast
import androidx.core.content.ContextCompat
import com.google.android.material.snackbar.Snackbar
import igrek.todotree.R
import igrek.todotree.info.logger.LoggerFactory
import igrek.todotree.ui.GUI
import igrek.todotree.ui.errorcheck.SafeClickListener
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.util.*

open class UserInfoService(private val activity: Activity, private val gui: GUI) {
    private val mainView: View? = null
    private val infobars = HashMap<View?, Snackbar>()
    private val logger = LoggerFactory.logger

    private fun resString(resourceId: Int): String {
        return activity.resources.getString(resourceId)
    }

    /**
     * @param info       tekst do wyświetlenia lub zmiany
     * @param view       widok, na którym ma zostać wyświetlony tekst
     * @param actionName tekst przycisku akcji (jeśli null - brak przycisku akcji)
     * @param action     akcja kliknięcia przycisku (jeśli null - schowanie wyświetlanego tekstu)
     */
    private fun showActionInfo(info: String, view: View, actionName: String?, action: InfoBarClickAction?, color: Int?) {
        var view: View? = view
        var action = action
        if (view == null) {
            view = mainView
        }
        var snackbar = infobars[view]
        if (snackbar == null || !snackbar.isShown) { //nowy
            snackbar = Snackbar.make(view!!, info, Snackbar.LENGTH_SHORT)
            snackbar.setActionTextColor(Color.WHITE)
        } else { //widoczny - użyty kolejny raz
            snackbar.setText(info)
        }
        if (actionName != null) {
            if (action == null) {
                val finalSnackbar: Snackbar = snackbar
                action = InfoBarClickAction { finalSnackbar.dismiss() }
            }
            val finalAction: InfoBarClickAction = action
            snackbar.setAction(actionName, object : SafeClickListener() {
                override fun onClick() {
                    finalAction.onClick()
                }
            })
            if (color != null) {
                snackbar.setActionTextColor(color)
            }
        }
        snackbar.show()
        infobars[view] = snackbar
        logger.info(info)
    }

    open fun showInfo(info: String) {
        showActionInfo(info, gui.mainContent, "OK", null, null)
    }

    open fun showInfoCancellable(info: String, cancelCallback: InfoBarClickAction?) {
        showActionInfo(info, gui.mainContent, "Undo", cancelCallback, ContextCompat.getColor(activity, R.color.colorPrimary))
    }

    fun showToast(message: String) {
        GlobalScope.launch(Dispatchers.Main) {
            Toast.makeText(activity.applicationContext, message, Toast.LENGTH_LONG).show()
        }
        logger.debug("UI: toast: $message")
    }

}